src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs

```diff
@@ -0,0 +1,216 @@
+
+            await Task.WhenAll(extractArtifactTasks);
+
+            // Delete the downloaded artifacts
+            Directory.Delete(inProcArtifactDir, true);
+            Directory.Delete(coreToolsHostArtifactDir, true);
+        }
+
+        private static string CreateStagingDirectory(string rootWorkingDirectory)
+        {
+            var stagingDirectory = Path.Combine(rootWorkingDirectory, StagingDirName);
+
+            if (Directory.Exists(stagingDirectory))
+            {
+                Directory.Delete(stagingDirectory, true);
+            }
+
+            Directory.CreateDirectory(stagingDirectory);
+            Console.WriteLine($"Created staging directory: {stagingDirectory}");
+
+            return stagingDirectory;
+        }
+
+        private static async Task<string> ExtractArtifactZipAsync(string artifactZipPath, string destinationDirectory)
+        {
+            if (!File.Exists(artifactZipPath))
+            {
+                throw new FileNotFoundException($"Artifact zip '{artifactZipPath}' does not exist.");
+            }
+
+            var extractedDir = FileUtilities.ExtractToDirectory(artifactZipPath, destinationDirectory);
+
+            // Artifact zip downloaded in the release task may contain multiple child zips. Extract them all.
+            await ExtractZipFilesInDirectoryAsync(extractedDir, extractedDir);
+
+            return extractedDir;
+        }
+
+        private async Task CreateCustomHostArtifactsAsync()
+        {
+            // Create a directory to store the assembled artifacts.
+            var customHostTargetArtifactDir = Path.Combine(_stagingDirectory, "artifacts-with-host");
+            Directory.CreateDirectory(customHostTargetArtifactDir);
+
+            var packTasks = _customHostTargetArtifactZipNames.Select(async artifactName =>
+            {
+                var runtimeDir = Directory.EnumerateDirectories(_inProc8ExtractedRootDir)
+                                          .FirstOrDefault(dir => dir.Contains(artifactName));
+                if (runtimeDir == null)
+                {
+                    throw new InvalidOperationException($"Runtime directory not found for {artifactName}");
+                }
+
+                var targetRuntimeName = Path.GetFileName(runtimeDir);
+                var targetRuntimeDir = Path.Combine(customHostTargetArtifactDir, targetRuntimeName);
+                Directory.CreateDirectory(targetRuntimeDir);
+
+                // Copy in-proc8 files
+                await Task.Run(() => FileUtilities.CopyDirectory(runtimeDir, Path.Combine(targetRuntimeDir, InProc8DirectoryName)));
+
+                // Copy in-proc6 files if exist
+                var inProc6Dir = Path.Combine(_inProc6ExtractedRootDir, targetRuntimeName);
+                if (Directory.Exists(inProc6Dir))
+                {
+                    await Task.Run(() => FileUtilities.CopyDirectory(inProc6Dir, Path.Combine(targetRuntimeDir, InProc6DirectoryName)));
+                }
```

Comment:
This feels surprising different from the .NET 8 version above. Should .NET 6 not also throw, for example? Would it make sense to encapsulate this logic and then just invoke it for both? The same question applies to a few places earlier on as well.

Reply:
Yes, both should throw. I fixed it now.